dm/master/shardddl/pessimist_test.go

```diff
@@ -279,3 +287,518 @@ func (t *testPessimist) TestPessimist(c *C) {
+	op13c.Done = true
+	done, _, err = pessimism.PutOperationDeleteExistInfo(etcdTestCli, op13c, i13)
+	c.Assert(err, IsNil)
+	c.Assert(done, IsTrue)
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		_, ok := p.Locks()[ID]
+		return !ok
+	}), IsTrue)
+	t.noLockExist(c, p)
+}
+
+func (t *testPessimist) TestUnlockSourceLackBeforeSynced(c *C) {
+	// some sources may be deleted (lack) before the lock become synced.
+	defer clearTestInfoOperation(c)
+
+	oriUnlockWaitOwnerInterval := unlockWaitInterval
+	unlockWaitInterval = 100 * time.Millisecond
+	defer func() {
+		unlockWaitInterval = oriUnlockWaitOwnerInterval
+	}()
+
+	var (
+		watchTimeout  = 500 * time.Millisecond
+		task          = "task-unlock-source-lack-before-synced"
+		source1       = "mysql-replica-1"
+		source2       = "mysql-replica-2"
+		source3       = "mysql-replica-3"
+		schema, table = "foo", "bar"
+		DDLs          = []string{"ALTER TABLE bar ADD COLUMN c1 INT"}
+		ID            = fmt.Sprintf("%s-`%s`.`%s`", task, schema, table)
+		i11           = pessimism.NewInfo(task, source1, schema, table, DDLs)
+		i12           = pessimism.NewInfo(task, source2, schema, table, DDLs)
+
+		sources = func(task string) []string {
+			switch task {
+			case task:
+				return []string{source1, source2, source3}
+			default:
+				c.Fatalf("unsupported task %s", task)
+			}
+			return []string{}
+		}
+		logger = log.L()
+		p      = NewPessimist(&logger, sources)
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// 0. start the pessimist.
+	c.Assert(p.Start(ctx, etcdTestCli), IsNil)
+	c.Assert(p.Locks(), HasLen, 0)
+	defer p.Close()
+
+	// no lock need to be unlock now.
+	c.Assert(terror.ErrMasterLockNotFound.Equal(p.UnlockLock(ctx, ID, "", false)), IsTrue)
+
+	// 1. PUT i11 & i12, will create a lock but now synced.
+	// not PUT info for source3 to simulate the deletion of it.
+	_, err := pessimism.PutInfo(etcdTestCli, i11)
+	c.Assert(err, IsNil)
+	rev1, err := pessimism.PutInfo(etcdTestCli, i12)
+	c.Assert(err, IsNil)
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		if len(p.Locks()) != 1 {
+			return false
+		}
+		_, remain := p.Locks()[ID].IsSynced()
+		return remain == 1
+	}), IsTrue)
+	c.Assert(p.Locks(), HasKey, ID)
+	synced, _ := p.Locks()[ID].IsSynced()
+	c.Assert(synced, IsFalse)
+	ready := p.Locks()[ID].Ready()
+	c.Assert(ready, HasLen, 3)
+	c.Assert(ready[source1], IsTrue)
+	c.Assert(ready[source2], IsTrue)
+	c.Assert(ready[source3], IsFalse)
+
+	// 2. try to unlock the lock manually, but the owner has not done the operation.
+	// this will put `exec` operation for the done.
+	c.Assert(terror.ErrMasterOwnerExecDDL.Equal(p.UnlockLock(ctx, ID, "", false)), IsTrue)
+
+	// 3. try to unlock the lock manually, and the owner done the operation.
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		// put done for the owner.
+		t.putDoneForSource(ctx, task, source1, i11, true, rev1, watchTimeout, c)
```

Comment:
Should we use `rev1 + 1`? Although it won't affect the result because we watch operations not infos

Reply:
changed to use `rev1 + 1` in https://github.com/pingcap/dm/pull/522/commits/03479c5d62477e67593019b51810a60ded21f9d2.